cpp/src/arrow/util/bit-util.cc

```diff
@@ -109,9 +109,37 @@ Status CopyBitmap(MemoryPool* pool, const uint8_t* data, int64_t offset, int64_t
   std::shared_ptr<Buffer> buffer;
   RETURN_NOT_OK(GetEmptyBitmap(pool, length, &buffer));
   uint8_t* dest = buffer->mutable_data();
-  for (int64_t i = 0; i < length; ++i) {
-    BitUtil::SetBitTo(dest, i, BitUtil::GetBit(data, i + offset));
+
+  int64_t byte_offset = offset / 8;
+  int64_t bit_offset = offset % 8;
+  int64_t num_bytes = BitUtil::BytesForBits(length);
+  int64_t bits_to_zero = num_bytes * 8 - length;
+
+  if (bit_offset > 0) {
+    auto carry_mask = BitUtil::kBitmask[bit_offset] - 1U;
+    auto carry_shift = 8U - bit_offset;
+
+    auto carry = 0U;
+    if (BitUtil::BytesForBits(length + bit_offset) > num_bytes) {
+      carry = (data[byte_offset + num_bytes] & carry_mask) << carry_shift;
+    }
+
+    auto i = num_bytes - 1;
+    while (i + 1 > 0) {
+      auto cur_byte = data[byte_offset + i];
```

Comment:
We should write down explicit types for these autos

Reply:
Replaced the autos with types.